### ISSUE TITLE
Const-eval errors in dead functions are optimization-dependent

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust#107503]
I tried this code: [[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1ffcf365aa942927bfb5f465c4dc4e99)

```rust
#![feature(inline_const)]

pub fn cause_late_error<T>() {
    const {
        panic!()
    };
}

fn main() {
    let a = 1;
    let b = 2;
    if a + b == 5 {
        cause_late_error::<()>();
    }
}
```

I expected to see this happen:

```text
error[E0080]: evaluation of `cause_late_error::<()>::{constant#0}` failed
 --> src/main.rs:7:9
  |
7 |         panic!()
  |         ^^^^^^^^ the evaluated program panicked at 'explicit panic', src/main.rs:7:9
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn cause_late_error::<()>`
  --> src/main.rs:13:9
   |
13 |         cause_late_error::<()>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
```

Instead, this happened: it compiled without any warnings.

### Meta

Playground nightly, 1.69.0-nightly (2023-01-30 001a77fac33f6560ff36)

### Related

- https://github.com/rust-lang/rust/issues/85836
- https://github.com/rust-lang/rust/pull/104087
- https://github.com/rust-lang/rust/issues/106617

Potential fix:

- #122568

**Comment by joshtriplett** [comment]
`let _: c_void = screaming.into();`

**Comment by oskgo** [comment]
On a related note, there's also this following code that compiles in release mode but fails to compile in debug mode: [[playground]](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=2e76b92776d5d72ac8f354131b45f88e)
```rust
#![feature(inline_const)]

#[inline(always)]
fn falser() -> bool {
    false
}

fn has_late_error<T>() {
    const {
        panic!()
    };
}

fn main() {
    if falser() {
        has_late_error::<()>();
    }
}
```
Maybe it makes more sense to track `inline_const`'s interactions with optimizations as a whole?

**Comment by CAD97** [comment]
I believe so, yes; any post-mono error should work. It's just significantly easier to show with panicking inline const than any other post-mono error, and much more accessible.

I believe this just results from `cause_late_error::<()>` being determined dead by MIR analysis and thus not monomorphized. Monomorphizing all mentioned names (enough to flush middleend post-mono errors) *should* fundamentally resolve the issue of constant evaluation errors disappearing because of MIR-level dead code elimination.

**Comment by RalfJung** [comment]
@joshtriplett yes.

```rust
struct S;

impl S {
  const C: () = panic!();
}

pub fn cause_late_error<T>() {
    S::C;
}

fn main() {
    let a = 1;
    let b = 2;
    if a + b == 5 {
        cause_late_error::<()>();
    }
}
```

**Comment by RalfJung** [comment]
And of course that also works for @oskgo's example:
```rust
#[inline(always)]
fn falser() -> bool {
    false
}

pub fn has_late_error<T>() {
    struct S;

    impl S {
      const C: () = panic!();
    }

    S::C;
}

fn main() {
    if falser() {
        has_late_error::<()>();
    }
}
```
Compiles in release mode but not in debug mode.

**Comment by zirconium-n** [comment]
Sorry if this sounds rude, but could we just issue a crater run and push a breaking change at this point? Panic in consteval is not something stabilized for too long. We can recognize this as a design mistake. Whether the code depends on something other than `#[cfg(debug_assertions)]` is a bit too far for me...

**Comment by RalfJung** [comment]
This is the same underlying issue as https://github.com/rust-lang/rust/issues/106617, so we probably should close one of these in favor of the other (or open a new issue to track the collected problems). We don't monomorphize dead code, which means
- since optimizations affect which code is considered dead, we do opt-dependent monomorphization
- since some errors are only raised during monomorphization, we have missing errors in dead code and opt-dependent errors

**Comment by traviscross** [comment]
This was mentioned over in https://github.com/rust-lang/rust/issues/106617#issuecomment-1424497797, but it's worth highlighting here that this same problem can be triggered without using `const` at all:

```rust
fn cause_late_error<T>() {
    cause_late_error::<((), T)>();
}

fn main() {
    let (a, b) = (1, 2);
    if a + b == 5 {
        cause_late_error::<()>();
    }
}
```

[Playground link](https://play.rust-lang.org/?version=beta&mode=release&edition=2021&gist=ab9768119b9b2c2665a8559ad3ea8e2a)

**Comment by RalfJung** [comment]
I think there's a fairly obvious way to fix this, if someone wants to give it a shot: we basically mirror `required_consts`.
- Add a new field `required_fns` to `mir::Body`
- Extend the `RequiredConstsVisitor` to also collect required functions (the visitor needs a rename then)
- When monomorphizing a function, we don't even have to walk the MIR body to find out which other functions need to be monomorhized recursively. We just monomorhize everything in `required_fn` (after applying suitable substitutions).

The big question is what the perf impact of that will be, but the only way to figure that out is to actually implement this. Any takers? :)

**Comment by oli-obk** [comment]
> Any takers? :)

on it

**Comment by CAD97** [comment]
Status:

- #112879 tried the straightforward approach (adding "required consts" to the "required types" collection that prevents this for type WF) but causes an unacceptably large perf regression (avg ~8%, up to 40% in primary) for unclear reason.
- #122568 takes a different approach which causes a much smaller regression (avg ~1%, up to 8%)

Some amount of compile perf regression is to be expected, since fixing this fundamentally requires evaluating more constants, and doing so takes time.

**Comment by RalfJung** [comment]
For the second one, where are you getting "up to 8%" from? [This](https://github.com/rust-lang/rust/pull/122568#issuecomment-2004680405) is the latest perf run, the worst regression in a primary benchmark is 2.4%. In secondary it's 3.8%. The average across all primary benchmarks (not just the ones that changed) is +0.33%. This is almost entirely due to regressions in incr builds; the worst affected category among primary benchmarks is incr-unchanged with +0.67%.

**Comment by RalfJung** [comment]
> Some amount of compile perf regression is to be expected, since fixing this fundamentally requires evaluating more constants, and doing so takes time.

Almost all the slowdown (around 2% of the 2.4%) occurs even if we just store the list of mentioned items, without changing anything about monomorphization. So almost the entire cost here is due to having to store this list as part of the crate metadata.

**Comment by CAD97** [comment]
> For the second one

Just clicked the wrong link to grab the stat, whoops (it's a bit harder on mobile to keep tabs open to constantly cross-reference). Fixed.

**Comment by RalfJung** [comment]
`g` should be a monomorphization root, if this is a library crate? We have a bunch of tests in https://github.com/rust-lang/rust/pull/122568 that look very similar and they all behave as intended.

**Comment by RalfJung** [comment]
Hm, is this related to automatically marking functions `#[inline]`? That would affect `should_codegen_locally`; I don't think it would affect reachability (so it shouldn't affect the roots) but I may miss something.

Does auto-`#[inline]` depend on the opt-level? That would be a problem. Cc @saethlin

**Comment by RalfJung** [comment]
Yes looks like the opt-level affects which functions are considered `#[inline]`

https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_mir_transform/src/cross_crate_inline.rs#L60-L66

Damn :/

**Comment by RalfJung** [comment]
What I don't understand is how making more things `#[inline]` makes fewer roots. I think it can only make strictly more things "reachable", so something about the later post-processing in "reachable_non_generic" must be removing inline functions. Maybe this?

https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L99

So, either we need to make the inline property opt-level-independent, or the collector has to gather some extra roots that are only used for "mentioned" processing.

**Comment by saethlin** [comment]
> Does auto-`#[inline]` depend on the opt-level? That would be a problem. Cc @saethlin

Doubly so, because a function is made inlinable if its `optimized_mir` contains no calls. If MIR optimizations are turned down or turned off, less functions are made inlinable because their calls aren't inlined away or otherwise optimized out.

**Comment by saethlin** [comment]
> What I don't understand is how making more things `#[inline]` makes fewer roots. I think it can only make strictly more things "reachable", so something about the later post-processing in "reachable_non_generic" must be removing inline functions. Maybe this?

It's this logic I think:
https://github.com/rust-lang/rust/blob/a7e4de13c1785819f4d61da41f6704ed69d5f203/compiler/rustc_passes/src/reachable.rs#L313-L324

I've found it much easier to think about these things in terms of a linker's view of the world. Reachable items are those which may be used by the linker. We never link to "inlined" (I really hate that term because it's used to mean so many things) items, which is why they're seen through but not included in reachability- they are codegenned in the caller crate so they make their contents visible to linkage but are not themselves linked against. That's why I've tried to push the "cross-crate" term, it makes more sense to me :shrug:

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Let's nominate to discuss this [recent observation](https://github.com/rust-lang/rust/issues/107503#issuecomment-2008127352) about the optimization-dependent behavior effect of the cross-crate inlining heuristic.

**Comment by RalfJung** [comment]
@saethlin that logic still keeps (cross-crate) inline functions in the reachable set though. In fact that logic makes *more* things reachable when a function is inline, since it walks the body. (Everything in the worklist gets added to reachable_symbols, at least for a lib crate as in @tmiasko's example.)

The issue is that the collector's roots aren't just everything reachable, it's a further restriction on this set:

https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_monomorphize/src/collector.rs#L1294-L1308

The compiler has at least 3 different notions of "reachable", and here the `MonoItemCollectionMode::Lazy` arm defines the collector idea of reachable, which involves `is_reachable_non_generic`.  (The name "reachable_non_generic" is extremely misleading. It's not just the reachable set reduced to non-generic items. It's a form of "codegen_reachable". But then the collector considers some things reachable that are not codegen-reachable. It's such a mess.) That is defined in the codegen backend via some "export threshold" stuff that I hope I can ignore

https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L156-L168

and then finally ends up in [this function](https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L45) which takes `reachable_set` and then removes a bunch of stuff. In particular

https://github.com/rust-lang/rust/blob/a8f71cf2893e03ec0ef7663f680f5d816e80f801/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L96-L105

So `generates_cgu_internal_copy` means we don't consider it a collector root, and that function finally [calls `cross_crate_inlinable`](https://github.com/rust-lang/rust/blob/7e1969ac13b94694bb6c9315dcecdc6b0ee344ad/compiler/rustc_middle/src/ty/instance.rs#L287).

---

So how could this be fixed?
I *think* the transitive closure of "functions mentioned in `reachable_set` " is optimization-independent. That set is bigger when more things can get inlined but it is only bigger by including more things contained in the body -- which are things that the collector would traverse anyway. (However it's not exactly the same kind of traversal so there is some chance of mismatch here.) So the collector could add items which are reachable but not codegen-reachable as "mentioned but not used" to the root set and then maybe that would make it independent of inlining heuristics?

I won't have time to experiment with this though for at least 2 weeks.

**Comment by RalfJung** [comment]
Btw there's also this gem

https://github.com/rust-lang/rust/blob/89b536dbc85e23275f08feaf73a7abce691bf72c/compiler/rustc_passes/src/reachable.rs#L430-L437

The entire handling of inlineable functions in reachable is very strange. We prune them from "codegen_reachable" and so we have to add more things eagerly to "reachable" so that they don't get entirely skipped by the collector...

I think conceptually what we actually want is 
- Start with a reasonable set of reachable roots -- public functions, lang items; basically the stuff that `reachable_set` uses to seed the search. That's the "externally reachable" set. Inlining is irrelevant. This is the only thing done pre-mono.
- Then later during monomorphization, build the full graph of *monomorphized* items starting from the non-generic items in that "externally reachable" set and traversing which other items are used. (There are actually two kinds of edges here, "used" and "mentioned", to deal with dead code optimized out.) Still inlining plays no role.
- Finally, the set of codegen'd functions is computed on that graph by starting with all non-inline functions that are "used"-reachable from an externally reachable item, and taking the transitive closure of "used" edges from there.

Currently we are trying to do the last step (finding all relevant non-inline functions) *before* monomorphization, leading to [some overapproximation](https://github.com/rust-lang/rust/blob/89b536dbc85e23275f08feaf73a7abce691bf72c/compiler/rustc_passes/src/reachable.rs#L430-L437) and an overall pretty messy structure.

---

@tmiasko ah, that's a fun one. But as long as this is optimization-level-independent it's not a problem. https://github.com/rust-lang/rust/pull/122568 deliberately determines "mentioned" items right before MIR opts, and e.g. after drop elaboration. We don't have a clear definition of what we consider "mentioned" currently. (E.g., if I have a variable of type `T` but it always gets moved out, is its destructor "mentioned"? There is a Drop terminator with that type, but it gets removed by drop elaboration.) So I don't intend to guarantee that "all functions that syntactically appear in `f` are mentioned"; all we need is that the "mentioned" set is the same in debug and optimized builds.

**Comment by RalfJung** [comment]
> The approximation is inherent to the problem since it is non-computable precisely. You can view the role of a reachability pass as a form of item collection that explores all possible instantiations of generic code. This work cannot be done by a mono item collector.

I have sketched a computable strategy in my previous post. I don't see why it is necessary to explore "all possible instantiations". The way I view it, `reachable_set` contains some hacks that are needed because the collector itself doesn't know how to deal with inlining properly -- so codegen particularities like "if a function can be inlined and it calls private functions, those private functions must be considered reachable" leak into reachability computation.

Both mono item collection and `reachable_set` are reachability computations. Mono item collection has all the information it needs to compute the precise set. But unfortunately its search gets pruned eagerly by removing inline functions too early, which then means some earlier pass has to *extend* the root set to ensure we don't remove too many things.

> Note that the example from https://github.com/rust-lang/rust/issues/107503#issuecomment-2008972955 is optimization level dependent.

I can't reproduce that, [this example](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=57c0fe9f886a7e5c613abeb3111c7e01) fails both in debug and release builds.

**Comment by RalfJung** [comment]
In general this behavior of skipping a bunch of checking on dead private functions is usually quite surprising anyway. Maybe the collector should just always walk *all* monomorphic functions, reachable or not, in "mentioned" mode?

**Comment by RalfJung** [comment]
> The example below fails only with optimizations enabled:

Oh lol so this is the other way around then, the unoptimized build misses the const-eval. Wtf, how is that happening...

**Comment by saethlin** [comment]
> Though I guess that means it has to run MIR opts on them, which is work we currently avoid doing. How likely is it that people have large amounts of monomorphic functions that never get mentioned (transitively) in anything "reachable"...?

FYI, https://github.com/rust-lang/rust/pull/109247 broke a handful of crates for example https://github.com/rust-lang/rust/issues/110475 because of invalid dead code. Errors in _monomorphic_ dead code I'm not sure, but errors in dead code surely exist, and I find it concerning that when we uncover these errors they get reported as regressions. Perhaps people would be more accepting of such a change if we announce it beforehand and expand our detection of erroneous code only once.

**Comment by RalfJung** [comment]
Ah, good point. I was mostly worried about performance. I don't know if there are cases where code would previously build with with and without optimizations, and then after that change it now fails in both cases.  But it's very possible.

I'm currently running this as a perf experiment in https://github.com/rust-lang/rust/pull/122568, but I don't intend to land it in the same change. The rest of https://github.com/rust-lang/rust/pull/122568 I am fairly sure only affects code that already fails when built without optimizations.

**Comment by RalfJung** [comment]
> Regarding whether this should block the stabilization of inline const, we tentatively felt that it should not, but we wanted to discuss that with more of the team before resolving the https://github.com/rust-lang/rust/pull/104087#issuecomment-1449080210 on https://github.com/rust-lang/rust/pull/104087.

Ah, reading that concern now gives me flashbacks. ;)

That concern makes a difference between "Optimization-dependent errors" and "Errors in unused code". It turns out those cannot be disentangled, as basically all notions of "used" that exist in rustc depend on optimization levels. (This was not the case back when the comment was written, but is true ever since `#[inline]` is automatically inferred and that inference depends on optimization levels.)

**Comment by saethlin** [comment]
The only way I can imagine making `cross_crate_inlinable` not optimization-dependent is by analyzing unoptimized MIR and doing the analysis at all optimization levels. If we do that, I am confident that we will regress both our compile time and our code generation quality (we could of course delete it altogether to avoid regressing compile time). But such a change does nothing to fix the behavior of `#[inline]`, which is the root (heh) cause here. I do not think that library authors would not be any more troubled by not getting errors due to `cross_crate_inlinable` than they would be by realizing that their inline-const assert that should have prevented shipping a bug did not do its job because they followed a clippy lint to put `#[inline]` on all their public items.

On the other hand, the fact that this bug has been shipping on nightly for quite some time and nobody has complained perhaps indicates that the impact is not too widespread.

I am very happy to see anyone trying to untangle the reachability/MIR collection code.

**Comment by RalfJung** [comment]
>  I do not think that library authors would not be any more troubled by not getting errors 

That was three "not" in a row, you lost me here. ;)

> I am very happy to see anyone trying to untangle the reachability/MIR collection code.

I gave up the hope of trying to simplify it, but maybe I can at least document it a bit better. ;)

**Comment by saethlin** [comment]
> That was three "not" in a row, you lost me here. ;)

I think library authors will be equally confused by `cross_crate_inlinable` implications as they will be by plain old `#[inline]` possibly hiding errors.

**Comment by RalfJung** [comment]
I have opened https://github.com/rust-lang/rust/issues/122814 for the collection root issue @tmiasko discovered.

**Comment by workingjubilee** [comment]
All uses in libraries I have seen are one of:
- top-level items
- only meant to stop you from _monomorphizing_ code with the "wrong" type

and I have certainly seen a few particularly... dank... usages of these kinds of "const assertions".

**Comment by RalfJung** [comment]
To be clear, the only blessed pattern libs may use is that if a const is directly used inside some function, then whenever that function executes with a given choice of its generic parameters, the const succeeded evaluating. So we do *not* guarantee that the code in the OP will fail to build!

The point of this issue is that we don't want optimization-level-dependent behavior. But which functions we walk and evaluate consts for is still not a stable guarantee -- except that any function that actually gets executed has been walked and had its consts evaluated.

**Comment by workingjubilee** [comment]
yeah, that seems fine? weird but fine.

**Comment by workingjubilee** [comment]
the OP's code fails to build in both debug and release now. huh. weird.

**Comment by RalfJung** [comment]
Probably needs an inline(never).
We have tests in tree that demo the opt dependent behavior; see the PR that closes this issue.

**Comment by RalfJung** [comment]
Hm, I wonder if inlining can also cause trouble here... if a call like `f::<i32>()` gets inlined, previously generic consts inside that body become monomorphic, and so if they ever get evaluated by later MIR passes then maybe that is evaluation that would not have happened otherwise (if this is all in a function that never gets collected)?

I'm beginning to wonder if MIR passes need a way to eval consts such that even if it fails, no error is emitted.

**Comment by RalfJung** [comment]
(After some experimentation with inlining) I think we aren't even computing `optimized_mir` for functions that do not participate in collection... in that case we're good. :)
